src/plugins/stacked-barchart/src/StackedBarchartPanel.tsx

```diff
@@ -0,0 +1,264 @@
+import React, { useLayoutEffect, useMemo } from 'react';
+import ReactDOM from 'react-dom';
+import useMeasure from 'react-use-measure';
+import { PanelProps, classicColors } from '@grafana/data';
+import { useTheme } from '@grafana/ui';
+import { css } from '@emotion/css';
+import * as d3 from 'd3';
+
+import { BarDatum, StackedBarchartOptions } from './types';
+import { prepareGraphableFields, countDigits } from './utils';
+
+interface Props extends PanelProps<StackedBarchartOptions> {}
+
+export const StackedBarchartPanel: React.FC<Props> = ({ options, data, width, height }) => {
+  const theme = useTheme();
+  const graphRef = React.useRef(null);
+  const graphTooltipRef = React.useRef(null);
+  const [legendRef, legendMeasure] = useMeasure();
+  const legendHeight = options.showLegend ? legendMeasure.height : 0;
+  const svgHeight = height - legendHeight;
+
+  const { stacks, xDistinctGroups, yDistinctGroups, xMaxGroupCount, error } = useMemo(
+    () => prepareGraphableFields(data?.series, options),
+    [data, options]
+  );
+
+  const margin = {
+    top: 5,
+    right: 5,
+    bottom: 40,
+    left: 20 + 12 * countDigits(xMaxGroupCount),
+  };
+  const colorScale = d3.scaleOrdinal().domain(yDistinctGroups).range(classicColors);
+
+  useLayoutEffect(() => {
+    if (error) {
+      return;
+    }
+    const tooltip = d3.select(graphTooltipRef.current).style('display', 'none');
+    const svg = d3.select(graphRef.current);
+    svg.selectAll('*').remove();
+
+    // X axis
+    const xAxis = d3
+      .scaleBand()
+      .domain(xDistinctGroups)
+      .range([0, width - margin.right - margin.left])
+      .padding(0.2);
+
+    const xAxisGenerator = d3.axisBottom(xAxis);
+
+    if (options.isXContinuous) {
+      let maxBinsModulo = Math.ceil(xDistinctGroups.length / options.numberOfBins);
+      xAxisGenerator.tickValues(xDistinctGroups.filter((_, index) => index % maxBinsModulo === 0));
```

Comment:
IMO we should not display every ticks for small event group intervals, it's ~barely~ unreadable:

![image](https://user-images.githubusercontent.com/956157/132662182-9c3c36ba-7832-4bb4-9b33-c55fdcd46440.png)

Reply:
Currently, there are two "manual" solutions:
On the plugin side, the `Number of bins` option is controlling the (max) number of x labels.
![Screenshot from 2021-09-09 15-00-37](https://user-images.githubusercontent.com/32192091/132690758-ba3ebec2-87fe-4f69-a0c1-daa8e1e3d628.png)

On the query side, The `Event group interval` query template reduces the number of bars:
![Screenshot from 2021-09-09 15-10-51](https://user-images.githubusercontent.com/32192091/132692019-94f1c676-c87e-4fd4-9400-4f3898bdf114.png)

Would it be preferable to have an option to control the number of bins and bars automatically ?